async-queue/src/fifo.rs

```diff
@@ -0,0 +1,403 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use super::*;
+use fortanix_sgx_abi::{FifoDescriptor, Return, Usercall};
+use lazy_static::lazy_static;
+use std::collections::HashSet;
+use std::mem;
+use std::sync::atomic::{AtomicU64, AtomicUsize, Ordering};
+use std::sync::Mutex;
+
+impl<T: WithAtomicId> Fifo<T> {
+    /// panics if len is not a power of two.
+    pub fn new(len: usize) -> Self {
+        assert!(len.is_power_of_two(), "Fifo len should be a power of two");
+        let mut data = Vec::with_capacity(len);
+        data.resize_with(len, T::empty);
+        Self {
+            data: data.into_boxed_slice(),
+            offsets: Box::new(AtomicUsize::new(0)),
+        }
+    }
+
+    pub fn descriptor(&mut self) -> FifoDescriptor<T> {
+        FifoDescriptor {
+            data: self.data.as_mut().as_mut_ptr(),
+            len: self.data.len(),
+            offsets: self.offsets.as_ref() as _,
+        }
+    }
+
+    pub fn sender<S: Synchronizer>(&mut self, synchronizer: S) -> Sender<T, S> {
+        Sender::new(self.descriptor(), synchronizer)
+    }
+
+    /// Panics if there is an existing (sync or async) receiver for the same queue.
+    pub fn receiver<S: Synchronizer>(&mut self, synchronizer: S) -> Receiver<T, S> {
+        Receiver::new(self.descriptor(), synchronizer)
+    }
+
+    pub fn async_sender<S: AsyncSynchronizer>(&mut self, synchronizer: S) -> AsyncSender<T, S> {
+        AsyncSender::new(self.descriptor(), synchronizer)
+    }
+
+    /// Panics if there is an existing (sync or async) receiver for the same queue.
+    pub fn async_receiver<S: AsyncSynchronizer>(&mut self, synchronizer: S) -> AsyncReceiver<T, S> {
+        AsyncReceiver::new(self.descriptor(), synchronizer)
+    }
+}
+
+pub(crate) fn try_send_impl<T: WithAtomicId>(descriptor: &FifoDescriptor<T>, val: &T) -> Result</*wake up reader:*/ bool, TrySendError> {
+    let (new, was_empty) = loop {
+        // 1. Load the current offsets.
+        let current = unsafe {
+            let offsets = (*descriptor.offsets).load(Ordering::SeqCst);
+            Offsets::new(offsets, descriptor.len as u32)
+        };
+        let was_empty = current.is_empty();
+
+        // 2. If the queue is full, wait, then go to step 1.
+        if current.is_full() {
+            return Err(TrySendError::QueueFull);
+        }
+
+        // 3. Add 1 to the write offset and do an atomic compare-and-swap (CAS)
+        //    with the current offsets. If the CAS was not succesful, go to step 1.
+        let new = current.increment_write_offset();
+        let current = current.as_usize();
+        let prev = unsafe {
+            (*descriptor.offsets).compare_and_swap(current, new.as_usize(), Ordering::SeqCst)
+        };
+        if prev == current {
+            break (new, was_empty);
+        }
+    };
+
+    // 4. Write the data, then the `id`.
+    let slot = unsafe { &mut *descriptor.data.add(new.write_offset()) };
```

Comment:
The safety requirements of `pointer::offset` and `pointer::add` are exactly the same. In fact:
> `pub unsafe fn add(self, count: usize) -> *mut T`
> Calculates the offset from a pointer (**convenience for .offset(count as isize)**).

Since in this case offsets are always 32 bits long, the cast to isize does not add any additional concern either.